ngcore/ngcore.go

```diff
@@ -104,79 +105,51 @@ func (n *Neugram) GetOrNewSession(ctx context.Context, name string) *Session {
 	return s
 }
 
-func (s *Session) Exec(src []byte) error {
+// Exec returns the evaluation of the content of src and an error, if any.
+// If src contains multiple statements, Exec returns the value of the last one.
+func (s *Session) Exec(src []byte) ([]reflect.Value, error) {
 	var err error
 	stdout := s.Stdout
 	if stdout == nil {
 		stdout, err = os.Create(os.DevNull)
 		if err != nil {
-			return err
+			return nil, err
 		}
 	}
 	stderr := s.Stderr
 	if stderr == nil {
 		stdout, err = os.Create(os.DevNull)
 		if err != nil {
-			return err
+			return nil, err
 		}
 	}
 
+	s.ExecCount++
+
 	res := s.Parser.ParseLine(src)
 	if len(res.Errs) > 0 {
 		errs := make([]error, len(res.Errs))
 		for i, err := range res.Errs {
 			errs[i] = err
 		}
-		return Error{Phase: "parser", List: errs}
+		return nil, Error{Phase: "parser", List: errs}
 	}
+	var out []reflect.Value
 	for _, stmt := range res.Stmts {
 		v, err := s.Program.Eval(stmt, nil)
 		if err != nil {
 			str := err.Error()
 			if strings.HasPrefix(str, "typecheck: ") { // TODO: gross
-				return Error{
+				return nil, Error{
 					Phase: "typecheck",
 					List: []error{
 						errors.New(strings.TrimPrefix(str, "typecheck: ")),
 					},
 				}
 			}
-			return Error{Phase: "eval", List: []error{err}}
-		}
-		s.ExecCount++
-		if len(v) > 1 {
-			fmt.Fprint(stdout, "(")
-		}
-		for i, val := range v {
-			if i > 0 {
-				fmt.Fprint(stdout, ", ")
-			}
-			if val == (reflect.Value{}) {
-				fmt.Fprint(stdout, "<nil>")
-				continue
-			}
-			switch v := val.Interface().(type) {
-			case eval.UntypedInt:
-				fmt.Fprint(stdout, v.String())
-			case eval.UntypedFloat:
-				fmt.Fprint(stdout, v.String())
-			case eval.UntypedComplex:
-				fmt.Fprint(stdout, v.String())
-			case eval.UntypedString:
-				fmt.Fprint(stdout, v.String)
-			case eval.UntypedRune:
-				fmt.Fprintf(stdout, "%v", v.Rune)
-			case eval.UntypedBool:
-				fmt.Fprint(stdout, v.Bool)
-			default:
-				fmt.Fprint(stdout, format.Debug(v))
-			}
-		}
-		if len(v) > 1 {
-			fmt.Fprintln(stdout, ")")
-		} else if len(v) == 1 {
-			fmt.Fprintln(stdout, "")
+			return nil, Error{Phase: "eval", List: []error{err}}
 		}
+		out = v
```

Comment:
Does this need to be an append if len(out) > 0?

Reply:
no, we piggy back on what IPython does: we just return the last statement.
(we can revisit afterwards, if needed)